File: Strohm/ios/StrohmNative/Sources/StrohmNative/Subscriptions.swift
import Foundation

class Subscriptions {
    let strohmNative: StrohmNative
    var pendingSubscriptions: [() -> Void]? = []
    var subscribers: [UUID: HandlerFunction] = [:]
    var subscribers2: [UUID: HandlerFunction2] = [:]

    init(strohmNative: StrohmNative) {
        self.strohmNative = strohmNative
        strohmNative.comms.registerHandlerFunction(name: "subscriptionUpdate",
                                                   function: self.subscriptionUpdateHandler)
    }

    func addSubscriber(propSpec: PropSpec,
                       handler: @escaping HandlerFunction,
                       completion: @escaping (UUID) -> Void,
                       onError: ErrorHandler?
    ) {
        if pendingSubscriptions != nil {
            pendingSubscriptions!.append({ [weak self] in
                self?.subscribe_(
                    propSpec: propSpec,
                    handler: handler,
                    completion: completion,
                    onError: onError
                )
            })
        } else {
            subscribe_(propSpec: propSpec,
                       handler: handler,
                       completion: completion,
                       onError: onError
            )
        }
    }

    private func subscribe_(propSpec: PropSpec,
                            handler: @escaping HandlerFunction,
                            completion: (UUID) -> Void,
                            onError: ErrorHandler?
    ) {
        guard let encodedPropSpec = strohmNative.comms.encode(object: propSpec) else {
            return
        }
        let stringEncoded = encodedPropSpec.replacingOccurrences(of: "\"", with: "\\\"")
        let subscriptionId = UUID()
        subscribers[subscriptionId] = handler
        strohmNative.call(
            method: "strohm_native.flow.subscribe_from_native(\"\(subscriptionId.uuidString)\", \"\(stringEncoded)\")",
            onError: onError
        )
        completion(subscriptionId)
    }

    func addSubscriber2(propSpec: PropSpec,
                        handler: @escaping HandlerFunction2,
                        completion: @escaping (UUID) -> Void,
                        onError: ErrorHandler?
    ) {
        if pendingSubscriptions != nil {
            pendingSubscriptions!.append({ [weak self] in
                self?.subscribe2_(
                    propSpec: propSpec,
                    handler: handler,
                    completion: completion,
                    onError: onError
                )
            })
        } else {
            subscribe2_(propSpec: propSpec,
                        handler: handler,
                        completion: completion,
                        onError: onError
            )
        }
    }

    private func subscribe2_(propSpec: PropSpec,
                             handler: @escaping HandlerFunction2,
                             completion: (UUID) -> Void,
                             onError: ErrorHandler?
    ) {
        guard let encodedPropSpec = strohmNative.comms.encode(object: propSpec) else {
            return
        }
        let stringEncoded = encodedPropSpec.replacingOccurrences(of: "\"", with: "\\\"")
        let subscriptionId = UUID()
        subscribers2[subscriptionId] = handler
        strohmNative.call(
            method: "strohm_native.flow.subscribe_from_native(\"\(subscriptionId.uuidString)\", \"\(stringEncoded)\")",
            onError: onError
        )
        completion(subscriptionId)
    }

    func removeSubscriber(subscriptionId: UUID) {
        subscribers.removeValue(forKey: subscriptionId)
        subscribers2.removeValue(forKey: subscriptionId)
        strohmNative.call(method: "strohm_native.flow.unsubscribe_from_native(\"\(subscriptionId.uuidString)\")") { cljsError in
            Log.warn("removeSubscriber failed with error \(cljsError)")
        }
    }

    func effectuatePendingSubscriptions() {
        if let pending = pendingSubscriptions {
            pendingSubscriptions = nil
            for subscriber in pending { subscriber() }
        }
    }

    func subscriptionUpdateHandler(args: CommsArguments) {
        if let subscriptionIdString = args["subscriptionId"] as? String,
           let subscriptionId = UUID(uuidString: subscriptionIdString),
           let newPropSerialized = args["new"] as? String {
            if let subscriber = subscribers[subscriptionId],
               let data = newPropSerialized.data(using: .utf8),
               let newProp = try? JSONSerialization.jsonObject(with: data) as? [Any],
               let propName = newProp.first as? String {
                subscriber((propName, newProp[1]))
            } else if let subscriber2 = subscribers2[subscriptionId] {
                subscriber2(newPropSerialized)
            }
        }
    }
}

File: Strohm/android/strohm-native/src/main/java/dev/strohmnative/Subscriptions.kt
package dev.strohmnative

import com.google.gson.Gson
import kotlinx.collections.immutable.PersistentList
import kotlinx.collections.immutable.persistentListOf
import java.util.UUID
import java.util.concurrent.atomic.AtomicMarkableReference

class Subscriptions {
    val strohmNative: StrohmNative

    internal var _pendingSubscriptions: AtomicMarkableReference<PersistentList<() -> Unit>>
    private var subscribers: MutableMap<UUID, HandlerFunction>

    internal constructor(strohmNative: StrohmNative) {
        this.strohmNative = strohmNative
        this._pendingSubscriptions = AtomicMarkableReference(persistentListOf(), true)
        this.subscribers = mutableMapOf()
        strohmNative.comms.registerHandlerFunction("subscriptionUpdate") { args ->
            this.subscriptionUpdateHandler(args)
        }
    }

    fun addSubscriber(
        propSpec: PropSpec,
        handler: HandlerFunction,
        completion: (UUID) -> Unit
    ) {
        var succeeded = false
        while (!succeeded) {
            val holder = BooleanArray(1)
            val pending = _pendingSubscriptions.get(holder)
            val isUsingPending = holder[0]

            if (isUsingPending) {
                val newPending = pending.add { this.subscribe_(propSpec, handler, completion) }
                succeeded = _pendingSubscriptions.compareAndSet(pending, newPending, true, true)
            } else {
                subscribe_(propSpec, handler, completion)
                succeeded = true
            }
        }
    }

    private fun subscribe_(
        propSpec: PropSpec,
        handler: HandlerFunction,
        completion: (UUID) -> Unit
    ) {
        val subscriptionId = UUID.randomUUID()
        subscribers[subscriptionId] = handler
        val encodedPropsSpec = strohmNative.comms.encode(propSpec)
        strohmNative.call("strohm_native.flow.subscribe_from_native(\"$subscriptionId\", \"$encodedPropsSpec\")")
        completion(subscriptionId)
    }

    internal fun removeSubscriber(subscriptionId: UUID) {
        subscribers.remove(subscriptionId)
        strohmNative.call("strohm_native.flow.unsubscribe_from_native(\"$subscriptionId\")")
    }

    internal fun effectuatePendingSubscriptions() {
        var isUsingPending: Boolean
        var pending: PersistentList<() -> Unit>
        var clearSucceeded: Boolean
        do {
            val holder = BooleanArray(1)
            pending = _pendingSubscriptions.get(holder)
            isUsingPending = holder[0]

            clearSucceeded = _pendingSubscriptions.compareAndSet(pending, persistentListOf(), true, false)
        } while (!clearSucceeded && isUsingPending)

        if (isUsingPending) {
            pending.forEach { it() }
        }
    }

    internal fun handlePropUpdate(prop: Prop, subscriptionId: UUID) {
        subscribers[subscriptionId]?.let { it(prop) }
    }

    internal fun subscriptionUpdateHandler(args: CommsHandlerArguments) {
        val subscriptionIdString = args["subscriptionId"] as? String ?: return

        val newEnc = args["new"] as? String
        val new = Gson().fromJson(newEnc, ArrayList::class.java)
        (new[0] as? PropName)?.let { propName ->
            @Suppress("UNCHECKED_CAST")
            val newProp = Prop(propName, new[1])
            val subscriptionId = UUID.fromString(subscriptionIdString)
            handlePropUpdate(newProp, subscriptionId)
        }
    }
}

